### ISSUE TITLE
Serialization

### ISSUE THREAD

**Issue by Psirus** [nutofem/nuto#61]
I know this is a tiring issue, but the state is annoying. It's broken so nobody uses it. The code is still there, cluttering up files and being adapted to new changes (at least sometimes), with no one being able to tell whether it will actually work. So maybe we can discuss a couple of questions here.

Is serialization useful? What use-cases can you imagine? How would you like the interface to be? Do we need to serialize everything? Will the work required to get it running again be justified by the usefulness of the feature?

**Comment by joergfunger** [comment]
I agree with this serialization issue being a never ending story that should be discussed with a rather high priority. The only reason for the introduction/keeping this serialization was the ability to write restart files and to split the visualization and the time integration. The serialization was then intended to be used for restoring the full structure (and then restoring the time dependent data separately for each time step) or even store the full structure as a binary (and restore the full structure for each time step).
### Pro
- Different formats can be used (text, binary, xml) and (at least in theory) only small additions in any header files are required to add a new class modify an existing class
- boost::serialize can automatically restore shared objects (via ptr/references), so when restoring those ptrs/references they are still pointing to the same location
- standard containers can automatically be serialized
### Con
- The implementation for many specific problems requires often a deeper understanding of what is going on in serialize (split of load and save; abstract classes; diamond structures; using templates, where the provided macros do not work; stack problems for nested fully connected structures ,...). Due to these problems, many people do not touch serialize at all
- Turning serialize on significantly increases the compile time - another reason for not using this in the development
- The intended visualization/restart option is not implemented yet
- The structure is not compatible with changes, thus just adding a new variable in any class (and serializing this) prevents a usage of an old serialization file (there are ways to do this, but keeping a compatible history is very cumbersome)

In my opinion, the restart/visualization issue will be urgent sooner or later, thus we somehow have to reconstruct the structure. There are two options I currently see
### Option 1
- using boost::serialize, but this has to be done in a joint effort and over a short period of time including tests and especially implementing the splitting between post-processing and time integration
### Option 2
- deleting boost::serialize from the source code. Restoring the (initial) structure is performed by just recreating the structure based on the mesh/constraints. This would require a general interface for the creation of the structure that has to be included in any main file 

~~~cpp
Structure& GenerateStructure()
{
    Structure myStructure(2);
    myStructure.ImportFromGmsh("ImportFile");
    myStrucure.ApplyConstraints ();
}
~~~
- When performing the post-processing, this methods have to be used to recreate the structure.
- Particular care has to taken to ensure the correct order (say the nodes on a ptr_container do not necessarily have the same order when being reconstructed) 
- In addition, we would need to (manually) write serialization routines for the nodal values, history data and time steps.
### Option 3
- The last option, is a combination of the two previous options. The reconstruction of the structure is done using option 2, the storage of the nodal values, and the integrands is done using the boost::serialize. However, this would require the store all the integrands in a serializeable container. In this option, the serilization functionality has to be included only the integrands (i.e. history/static data).

**Comment by TTitscher** [comment]
I like option 2 as a starting point, with a modification. IMO we do not need this general interface. We just re-run the main file until we reach TimeIntegration::Solve(). This will recreate the structure and its state, except time step, nodal values and history data. Those three things are already (manually) implemented, e.g. [here for the whole structure](https://github.com/nutofem/nuto/blob/master/src/mechanics/structures/StructureBase.cpp#L1027).

This approach has its limitations when it comes to changes of the _structure_ of the `Structure`, like mesh refinement or changes in the constraint equations. Maybe [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it). So I vote for removing boost::serialize.

And if we do need it again...
First, the current implementation is incomplete and requires touching any serialize code anyways. And secondly, a lot of the problems with boost::serialize are gone. A the cyclic dependencies including the structure are gone. There are no diamonds of deaths anymore.

**Comment by potto1** [comment]
The state we have reached some time ago was that serialization is actually not applicable to nuto because of the cyclic dependencies, so we had to implement a workaround (saving pointer addresses and updating them) in order to get it work. I guess in the actual nuto implementation some difficulties arose  in the former implementation may not come up (Diamond in NodeDof etc...), but the main problem would remain, so we need to apply the same workaround. From my point of view the PDE concept may change things significantly, and only after PDE implemented we have to think about this issue.
